File: 543-diameter-of-binary-tree/diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int diameterOfBinaryTree(TreeNode root) {
        if(root ==null) return 0;

        int leftdiam= diameterOfBinaryTree(root.left);
        int rightdiam= diameterOfBinaryTree(root.right);
        int currdiam= height(root.left)+height(root.right);

        return Math.max(Math.max(leftdiam,rightdiam),currdiam);
    }

    public int height(TreeNode root){
        if(root== null) return 0;
        int leftht = height(root.left);
        int rightht = height(root.right);

        return Math.max(leftht, rightht) + 1;

    }
}

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
int maxdiameter=0;
    int diameterOfBinaryTree(TreeNode* root) {
        height(root);
        return maxdiameter;
    }
    int height(TreeNode* node){
        if(node== NULL) return 0;
        int leftheight= height(node->left);
        int rightheight= height(node->right);
        maxdiameter= max(maxdiameter,leftheight+rightheight);

        return max(leftheight,rightheight)+1;
    }
};